### ISSUE TITLE
Changed receiver type and return lifetime

### ISSUE THREAD

**Issue by cuviper** [obi1kenobi/cargo-semver-checks#961]
### Is this about an existing lint, or proposing a new one?

new

### Known issues that might be causing this

- [ ] Is the item that should have been flagged originating from another crate, or from a re-export of such an item? ([#638](https://github.com/obi1kenobi/cargo-semver-checks/issues/638))
- [X] Is the issue related to a change in the type of a field, function parameter, generic, or trait bound? ([#149](https://github.com/obi1kenobi/cargo-semver-checks/issues/149))
- [ ] If you're proposing a new lint, is it already part of our lint wishlist? ([#5](https://github.com/obi1kenobi/cargo-semver-checks/issues/5))

### Steps to reproduce the bug with the above code

I was curious if semver-checks could detect this intentional breaking change:
https://github.com/indexmap-rs/indexmap/pull/304

```
$ git clone https://github.com/indexmap-rs/indexmap
$ cd indexmap
$ git checkout 244bd81
$ cargo semver-checks --baseline-rev @^
```

### Actual Behaviour

```
     Parsing indexmap v2.2.1 (current)
      Parsed [   2.700s] (current)
     Parsing indexmap v2.2.0 (baseline)
      Parsed [   2.695s] (baseline)
    Checking indexmap v2.2.0 -> v2.2.1 (patch change)
     Checked [   0.027s] 89 checks: 89 pass, 0 skip
     Summary no semver update required
```

### Expected Behaviour

I hoped it would detect the same change that `cargo public-api diff` finds:

```
$ cargo public-api diff @^..@
[...]
Changed items in the public API
===============================
-pub fn indexmap::map::raw_entry_v1::RawOccupiedEntryMut<'a, K, V, S>::into_key(&mut self) -> &mut K
+pub fn indexmap::map::raw_entry_v1::RawOccupiedEntryMut<'a, K, V, S>::into_key(self) -> &'a mut K
```

This could probably be two different lints, one for the changed receiver and one for the return lifetime.

#### Software version

cargo-semver-checks 0.35.0

#### Operating system

Linux 6.10.10-200.fc40.x86_64

#### Command-line

```bash
/home/jistone/.cargo/bin/cargo-semver-checks semver-checks --bugreport
```

#### cargo version

```
> cargo -V
cargo 1.81.0 (2dbb1af80 2024-08-20)
```

#### Compile time information

- Profile: release
- Target triple: x86_64-unknown-linux-gnu
- Family: unix
- OS: linux
- Architecture: x86_64
- Pointer width: 64
- Endian: little
- CPU features: fxsr,sse,sse2
- Host: x86_64-unknown-linux-gnu

**Comment by obi1kenobi** [comment]
Thanks! Agreed, probably two desirable lints here as you said: receiver changed + return lifetime change.

It's possible we might be able to catch these before we implement all of #149.

For the "receiver changed" lint, I'd like to make sure our approach is robust to the possible future changes from the "arbitrary receiver types" RFC. I'll need to think about the edge cases there.

**Comment by cuviper** [comment]
FWIW, I remembered there was also an intentional receiver change way back in https://github.com/rust-lang/rust/pull/39466, between Rust 1.15.0 and 1.15.1. Maybe these examples aren't particularly motivating since they were on purpose though. :)

**Comment by obi1kenobi** [comment]
The receiver changes are resolved as of #1222 and the new lints will be part of the next `cargo-semver-checks` release!

I'll keep this issue open since we can't lint for the return lifetime change yet.

**Comment by dharmvr1** [comment]
Hi @obi1kenobi ! I’d like to take a shot at the remaining part of this ,
My rough plan is to copy the pattern from the existing receiver lints in src/lints/*.ron and make a new lint that compares baseline vs current return lifetime (conservative first version: only obvious breaking cases). I’ll add a small baseline/current test crate + a test in src/query.rs under tests_lints, then regenerate rustdocs.
Quick question: do you know which schema field/path is best for extracting the return lifetime from rustdoc JSON? Also should I focus on inherent methods first or include trait methods too?
If this sounds fine I’ll start working and open a PR.

**Comment by dharmvr1** [comment]
Hey @obi1kenobi , I am exploring  Trustfall schema in trustfall-rustdoc-adapter. return_value maps to ReturnValue, and ReturnValue currently only exposes is_unit. So it seems return lifetime changes can’t be linted without first extending the adapter/schema to expose return type/lifetime info.
If that’s correct, I can try to work on exposing return type (maybe as a rendered string first), and then add the lint here.

**Comment by obi1kenobi** [comment]
Hi! Thanks for your interest in this open issue.

> So it seems return lifetime changes can’t be linted without first extending the adapter/schema to expose return type/lifetime info.

This is correct. Unfortunately even exposing that info isn't sufficient though, because we also need lifetime variance information and "outlives" bounds etc. Some of that isn't available in rustdoc JSON itself yet, and even if it were, we haven't worked out the precise rules around lifetime changes with variance in mind yet.

All that's to say, this is a project that is likely going to require months of committed work to accomplish. Is that something you're looking to take on?

As a first project in this codebase, I usually recommend writing a lint from an issue marked `E-mentor`, which should be doable relatively quickly and would be a good jumping off point for future work in the project. But if you specifically want to jump directly into the deep end, I wouldn't stop you :)
